feat!: Add `AlterField` Operation to Migration Generator

kumarUjjawal opened the pull request:
## Add `AlterField` Operation to Migration Generator

## Overview

This PR introduces support for the `AlterField` operation in the migration generator. 

## Related Issues

#205 

## Reviewer Notes

- I don't have full confidence in my implementation if you think this is not what was the intention of the #205 feel free to close this for now.

kumarUjjawal committed:
feat: Add `AlterField` Operation to Migration Generator

kumarUjjawal commented:
@seqre any feedback for me?

seqre reviewed (changes requested):
This operation has many edge cases, and it should be thoroughly tested. Can you please add more tests first? 

Think about possible cases, such as: name change (this one can be skipped for now, it's tricky), type change, attributes change, etc. All those should be either handled or visibly err when a given case is unhandled or ambiguous. 

It seems that `cot-cli/src/migration_generator.rs` does not have a test for alter model, would you be so kind and add that too?

kumarUjjawal commented:
Sure, Thank you!

kumarUjjawal committed:
added some more tests

kumarUjjawal committed:
added missing docs

kumarUjjawal commented:
@m4tx the error doesn't seem to come from my PR, can you take a look?

seqre reviewed (commented):
Thank you for your patience. I fixed the PR so it compiles and the tests pass! I left two comments and I'll try to test it myself tonight.

kumarUjjawal committed:
Marked DynOperation as #[non_exhaustive]

seqre committed:
Revert "added missing docs"

This reverts commit 3d80f57b6c1392a494d5681eeaee2e8f7efb00a6.

seqre commented on cot/src/db/migrations.rs:
```diff
@@ -521,6 +526,17 @@ impl Operation {
+                    .modify_column(new_field.as_column_def(database))
+                    .to_owned();
+                database.execute_schema(query).await?;
+            }
```
Could you please write some tests that cover this part?

seqre commented on cot/src/db/migrations.rs:
```diff
@@ -594,6 +610,18 @@ impl Operation {
+                    .modify_column(old_field.as_column_def(database))
+                    .to_owned();
+                database.execute_schema(query).await?;
+            }
```
Could you please write some tests that cover this part?

seqre reviewed (changes requested):
Hey @kumarUjjawal, it seems you force-pushed your old changes which broke tests and removed the custom operations added recently to `master`. I've fixed those for you, please remember to `git pull` before doing any changes!

Regarding the PR, we're getting closer with every iteration! This feature is really important, but it's slightly complicated, that's why I'm pushing for it to be well tested. I appreciate your efforts!

seqre commented:
Hey @kumarUjjawal, how's the work going on this PR? Do you need any help from us?

kumarUjjawal commented:
Hi @seqre Sorry for the delay, I got busy with another project, doesn't seem like I can give time for the next few weeks Can you pick this up from here?

claude committed:
feat(migrations): address review feedback on AlterField

- Restore the "Modified" status message after constructing the
  AlterField operation in make_alter_field_operation, mirroring the
  Adding/Added pattern used by make_add_field_operation.
- Add rustdoc and examples to Operation::alter_field() and
  AlterFieldBuilder (and its methods), matching the style of
  remove_field/remove_model. Examples are marked no_run because SQLite
  does not support ALTER TABLE ... MODIFY COLUMN.
- Add a unit test (test_operation_alter_field) and PostgreSQL/MySQL
  database tests for AlterField forwards/backwards. SQLite is skipped
  because sea_query's SQLite backend panics on modify_column; the
  postgres/mysql tests follow the pattern used by other ignored
  postgres/mysql tests.
- Add unit tests covering the new AlterField branch in
  get_ops_adding_foreign_keys: one where the new field gains a foreign
  key and one where the new field has no foreign key.

seqre committed:
do not handle foreign key constraints yet

pre-commit-ci[bot] committed:
chore(pre-commit.ci): auto fixes from pre-commit hooks

m4tx committed:
Merge branch 'master' into alter-field

# Conflicts:
#	cot-cli/src/migration_generator.rs

m4tx committed:
fixup: adapt to master after merge

- cargo fmt formatting for alter_field doc example
- add on_delete/on_update to ForeignKeySpec test literals (added on
  master by the on_delete/on_update FK feature)

m4tx commented:
> Hi @seqre Sorry for the delay, I got busy with another project, doesn't seem like I can give time for the next few weeks Can you pick this up from here?

Hey @kumarUjjawal, sure, that's alright! I've tried to pick it up and found out that the implementation of this for SQLite will require some more complex work, as this database engine doesn't support the `ALTER FIELD` syntax. Other frameworks typically emulate this by a "create-copy-drop-rename" table dance, but it's not trivial to do this with Cot currently as it would require providing some more context to the migration engine.

I think we'll take it from here and will be able to finish this; thanks for laying out the groundworks for this!
